cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.sourcecontrol;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.sourcecontrol.CommitMeta;
+import io.cdap.cdap.sourcecontrol.RepositoryManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+public class InMemorySourceControlOperationRunner implements SourceControlOperationRunner {
+  private final RepositoryManager repositoryManager;
+  private final CConfiguration cConf;
+
+  private static final Logger LOG = LoggerFactory.getLogger(InMemorySourceControlOperationRunner.class);
+
+  @Inject
+  public InMemorySourceControlOperationRunner(CConfiguration cConf, RepositoryManager repositoryManager) {
+    this.cConf = cConf;
+    this.repositoryManager = repositoryManager;
+  }
+
+  @Override
+  public ListenableFuture<PushAppsResponse> push(List<AppDetailsToPush> appsToPush, CommitMeta commitDetails,
+                                                 @Nullable String branchName) throws PushFailureException {
+
+    try {
+      repositoryManager.startAndWait();
+      repositoryManager.switchToCleanBranch(branchName);
+    } catch (Exception e) {
+      LOG.debug("Failed to initialise repository: %s", e.getCause());
+      throw new PushFailureException("Failed to initialise repository", e);
+    }
+
+    Path repoBasePath = repositoryManager.getBasePath();
+
+    //TODO: handle if appsToPush is too large
+    LOG.info("Pushing application configs for : {}", appsToPush.stream().map(AppDetailsToPush::getApplicationName));
+
+    Map<String, Path> applicationPathMap =
+      appsToPush.stream().collect(Collectors.toMap(AppDetailsToPush::getApplicationName,
+                                                   appDetail -> repoBasePath.resolve(appDetail.getApplicationName())));
```

Comment:
Do we want to create a directory for each artifact type (eg: streaming, data pipeline)?

Reply:
Can you also make a method for getting the repo path based on appDetail? It will be used by pull also.